_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+
+Nonetheless, we quickly relate what was done, as this task is a nice example of handling data where features vary over more than one axis.
+
+Preprocessing starts as in the aforementioned previous posts. Here though, we explicitly use eager execution, and may sometimes hard-code known values to keep the code snippets short.
+
+```{r}
+library(tensorflow)
+library(tfprobability)
+
+tfe_enable_eager_execution(device_policy = "silent")
+
+library(tfdatasets)
+library(dplyr)
+library(readr)
+library(purrr)
+library(caret)
+library(stringr)
+
+# make decode_wav() run with the current release 1.13.1 as well as with the current master branch
+decode_wav <- function() if (reticulate::py_has_attr(tf, "audio")) tf$audio$decode_wav
+  else tf$contrib$framework$python$ops$audio_ops$decode_wav
+# same for stft()
+stft <- function() if (reticulate::py_has_attr(tf, "signal")) tf$signal$stft else tf$spectral$stft
+
+files <- fs::dir_ls(path = "audio/data_1/speech_commands_v0.01/", # replace by yours
+                    recursive = TRUE,
+                    glob = "*.wav")
+
+files <- files[!str_detect(files, "background_noise")]
+
+df <- tibble(
+  fname = files,
+  class = fname %>%
+    str_extract("v0.01/.*/") %>%
+    str_replace_all("v0.01/", "") %>%
+    str_replace_all("/", "")
+)
+```
+
+We train the MAF on pronunciations of the word "zero".
+
+```{r}
+for (c in unique(df$class)) {
+  assign(paste0("df_", c), df %>% filter(class == c) %>% select(fname))
+}
+
+df_ <- df_zero # 2 * 1178 rows
+idx_train <- sample(1:nrow(df_), 0.5 * nrow(df_))
+df_train <- df_[idx_train, ]
+df_test <- df_[-idx_train, ]
+```
+
+Following the approach detailed in [Audio classification with Keras: Looking closer at the non-deep learning parts](https://blogs.rstudio.com/tensorflow/posts/2019-02-07-audio-background/), we'd like to train the network on spectrograms instead of the raw time domain data. 
+Using the same settings for `frame_length` and `frame_step` of the Short Term Fourier Transform as in that post, we'd arrive at data shaped `number of frames x number of FFT coefficients`. To make this work with the `masked_dense()` employed in `tfb_masked_autoregressive_flow()`, the data would then have to be flattened, yielding an impressive 25186 features in the joint distribution.
+
+With the architecture defined as above in the GAS example, this lead to the network not making much progress. Neither did leaving the data in time domain form, with 16000 features in the joint distribution. Thus, we decided to work with the FFT coefficients computed over the complete window instead, resulting in 257 joint features.^[We quickly experimented with a higher number of FFT coefficients, but the approach did not seem that promising.]
+
+```{r}
+batch_size <- 100
+
+sampling_rate <- 16000L
+data_generator <- function(df,
```

Comment:
This function returns `NULL`, do we need to remove the last assignment?

Reply:
absolutely, right, too many versions ...